src/osd/OSD.cc

```diff
@@ -7834,6 +7834,10 @@ void OSD::_committed_osd_maps(epoch_t first, epoch_t last, MOSDMap *m)
       // set incarnation so that osd_reqid_t's we generate for our
       // objecter requests are unique across restarts.
       service.objecter->set_client_incarnation(osdmap->get_epoch());
+      // force re-sending any pending failure reports in case
+      // they might get ignored by monitor (e.g., because I was
+      // previously marked as dead)
+      requeue_failures();
```

Comment:
What? We deliberately *don't* requeue failure reports when transiting from dead to active. We assume that the dead OSD was in fact the one experiencing problems. Perhaps that assumption isn't *always* correct, but it's definitely *often* correct.

Reply:
The problem is that if a dead OSD is truly healthy again,  it won't be able to vote correctly until all existing pending failure reports get requeued (e.g., unless we are re-connecting to a new monitor) . Also it does not hurt to requeue failure reports here because if this OSD is still the one experiencing problems, its reports will simply get ignored again, and this OSD will be re-voted and re-marked as dead.

Reply:
So why not just reset the failure queues and let it gather new ones?

Reply:
We can not simply throw out all the old pending failures here because we still have to initiatively cancel any failure reports once the heartbeat connections between are restored, otherwise we'll leave a dirty pending failure history to the monitor side, which is generally error-prone.

Reply:
I don't understand what you mean here. Surely we can clean up the failure state (whether it's local or remote) without claiming to have detected failures that might be our fault?

Reply:
It is by design the responsibility for an OSD to send out failure reports on detecting heartbeat errors and to cancel if the corresponding broken heartbeat connections get fixed, and monitor relies on the total valid failure reports (e.g., reports from OSDs locating in different hosts) to make decisions and mark an pending failure OSD as down. Hence we should try our best to make the heartbeat behaviour to match the above constraint whenever possible.

The key point is that the entries in the ```failure_pending``` queue here represents OSDs we've sent out failure reports to monitor but are not being marked as down yet. We could re-queue, re-report, re-cancel, or whatever, which generally should not pose a problem. But I'd suggest not throwing out all the pending failures here, since we might lose the chance of fixing the false failures we've already sent out(e.g., we cleared the ```failure_pending``` queue, and we do not heartbeat with some of them anymore). On the other hand, by calling ```requeue_failures()``` here, we'll re-send all the failure reports in the ```failure_pending``` queue and we may still be able to cancel if any of the heartbeat connections turns out to be good now.

Reply:
Okay, I've refreshed myself on how the failure reporting to the monitor works again. It seems like this is a bit of a hole in the design: as you say, the monitors will just ignore received reports from any down OSDs, but they don't try and clean up existing reports when they mark an OSD down.

So my concern is that an OSD which has been marked down is definitely going to have heartbeat failures from well-behaving up OSDs, because those OSDs will have ignored the requests from the down one! But it may take them a little while to gather the newest OSD map and start responding to it again.

Just to enumerate some options:
1) do what this code does, and send in all our failure reports to the monitor again. Note that this might mean the monitor decides it has two failure reports on our peers when we really should only have sent one; conversely, if we do know about a dead peer it will get reported to the monitor that much faster.
2) Throw out all the pending failures. I think this is what the existing code intended to do, but wasn't handling because it didn't actually clean up the ```failure_pending``` queue?
3) Proactively cancel all our ```failure_pending``` reports to the monitor (by sending in a FLAG_ALIVE message), and then throw them out. This has the advantage of cleaning up any of our erroneous error reports quickly and not leaking the failures, but the disadvantage of slowing down detection of any true failures.
4) something else?

So I think false failures are a bit expensive and can help lead to runaway cluster thrashing situations, which is why I don't like choice (1). Choice (3) is probably pretty simple and addresses your cleanup concern and might be the best option — if an OSD is truly dead, it's not likely our solitary report will make a difference in any cluster of size; in small clusters it will still be a quick timeout before we report it dead.

Reply:
> Choice (3) is probably pretty simple and addresses your cleanup concern and might be the best option — if an OSD is truly dead, it's not likely our solitary report will make a difference in any cluster of size; in small clusters it will still be a quick timeout before we report it dead.

As you wish.